sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,595 @@
+    def omega(self):
+        """Set of live cosets
+        """
+        return [coset for coset in range(len(self.p)) if self.p[coset] == coset]
+
+    @property
+    def n(self):
+        """The number `n` represents the largest number that has been used so
+        far for a live coset
+        """
+        return max(self.omega) + 1
+
+    # checks whether the Coset Table is complete or not
+    def is_complete(self):
+        return not None in flatten(self.table)
+
+    # Pg. 153
+    def define(self, alpha, x):
+        A = self.A
+        if self.n == DefaultMaxLimit:
+            # abort the further generation of cosets
+            return
+        self.table.append([None]*len(A))
+        # beta is the new coset generated
+        beta = len(self.table) - 1
+        self.p.append(beta)
+        self.table[alpha][A.index(x)] = beta
+        self.table[beta][A.index(x.inverse())] = alpha
+
+    def define_f(self, alpha, x):
+        A = self.A
+        if self.n == DefaultMaxLimit:
+            # abort the further generation of cosets
+            return
+        self.table.append([None]*len(A))
+        # beta is the new coset generated
+        beta = len(self.table) - 1
+        self.p.append(beta)
+        self.table[alpha][A.index(x)] = beta
+        self.table[beta][A.index(x.inverse())] = alpha
+
+    def scan(self, alpha, word):
+        """
+        >>> from sympy.combinatorics.free_group import free_group
+        >>> from sympy.combinatorics.fp_groups import CosetTable, FpGroup
+        >>> F, x, y = free_group("x, y")
+        >>> f = FpGroup(F, [x**3, y**3, x**-1*y**-1*x*y])
+        >>> c = CosetTable(f, [x])
+
+        """
+        # alpha is an integer representing a "coset"
+        # since scanning can be in two cases
+        # 1. for alpha=0 and w in Y (i.e generating set of H)
+        # 2. alpha in omega (set of live cosets), w in R (relators)
+        f = alpha
+        i = 0
+        r = len(word)
+        # list of union of generators and their inverses
+        A_dict = self.A_dict
+        while i < r and self.table[f][A_dict[word.subword(i, i+1)]] is not None:
+            f = self.table[f][A_dict[word.subword(i, i+1)]]
+            i += 1
+        # can this be replaced with i == r ?
+        if i >= r:
+            if f != alpha:
+                # implement the "coincidence" routine on Pg 158 of Handbook.
+                self.coincidence(f, alpha)
+                return
+        b = alpha
+        j = r - 1
+        while j >= i and self.table[b][A_dict[word.subword(j, j+1)**-1]] is not None:
+            b = self.table[b][A_dict[word.subword(j, j+1).inverse()]]
+            j -= 1
+        if j < i:
+            # we have an incorrect completed scan with coincidence f ~ b
+            # run the "coincidence" routine
+            # line: We assume that before each call of COINCIDENCE, for α ∈ [1..n],
+            # we have p[α]=α iff α ∈ Ω.
+            self.p.extend(list(range(len(self))))
```

Comment:
Yes, I know this line is to be removed. The other method `scan_and_fill` doesn't have this line.